### PR TITLE
[#29] 1/2 - Prepare some necessary changes in the shared module before applying the Android submodule

### LINKED ISSUE #29: Reuse the "android" module from android-templates
## Why

In the KMM project structure, the "android" module acts nearly the same as a normal Android app module in the native Android project, (i.e., from [android-templates](https://github.com/nimblehq/android-templates/tree/develop/template-compose)). There are many same configs, structures, and dependencies that could be shared and reused from the [android-templates's app](https://github.com/nimblehq/android-templates/tree/develop/template-compose/app) for the KMM's "android" module.

To leverage and reuse the "android" app module from android-templates, we should consider resolving these issues:
- use `git sub-module` to fetch the app module from the [template-compose](https://github.com/nimblehq/android-templates/tree/develop/template-compose/app) of the android-templates.
- android-templates's app uses Hilt (Dagger) while KMM's android uses Koin.
- android-templates's app interacts with its data and domain modules, while KMM's android interacts with its data and domain from the "shared" module.
- android-templates's app uses Gradle 7.3 while KMM's android uses Gradle 8.
- and so on.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by luongvo**
## What happened 👀

- Fix the `make_ios` script to be re-run multiple times continuously.
- Refactor Android module dependencies in the `shared` module.
- Refactor component names (remove `User` from all components) & packages in the `shared` module.
- Rename Android's app id to `co.nimblehq.kmm.template`.

## Insight 📝

- Use the plural for packages with homogeneous contents and the singular for packages with heterogeneous contents. [softwareengineering.stackexchange.com/questions/75919/should-package-names-be-singular-or-plural](https://softwareengineering.stackexchange.com/questions/75919/should-package-names-be-singular-or-plural).
- The generated iOS source code should be cleaned up before executing the next run.

  image

## Proof Of Work 📹

- The iOS script is now able to be re-run multiple times continuously ✅ 

https://github.com/nimblehq/kmm-templates/assets/16315358/e5b02d2c-2a9e-44e8-8684-be1d566db04b

### CODE REVIEW

File: sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt [hunk 1]
```diff
@@ -13,33 +13,33 @@ import kotlin.test.BeforeTest
...
 
     @BeforeTest
     fun setUp() {
-        useCase = GetUsersUseCaseImpl(mockRepository)
+        useCase = UseCaseImpl(mockRepository)
     }
 
     @Test
-    fun `When calling getUsers successfully - it returns users`() = runTest {
+    fun `When calling successfully - it returns users`() = runTest {
```
Comment (kaungkhantsoe) [inline sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt:29]: This is still calling `getUsers`. Wouldn't it be better without changes?

File: sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt [hunk 2]
```diff
@@ -13,33 +13,33 @@ import kotlin.test.BeforeTest
...
         useCase().test {
-            awaitItem() shouldBe listOf(user)
+            awaitItem() shouldBe listOf(model)
             awaitComplete()
         }
     }
 
     @Test
-    fun `when calling getUsers fails - it throws the corresponding error`() = runTest {
+    fun `when calling fails - it throws the corresponding error`() = runTest {
```
Comment (kaungkhantsoe) [inline sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt:42]: Ditto

File: shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt [hunk 3]
```diff
@@ -13,33 +13,33 @@ import kotlin.test.BeforeTest
...
 
     @BeforeTest
     fun setUp() {
-        useCase = GetUsersUseCaseImpl(mockRepository)
+        useCase = UseCaseImpl(mockRepository)
     }
 
     @Test
-    fun `When calling getUsers successfully - it returns users`() = runTest {
+    fun `When calling successfully - it returns users`() = runTest {
```
Comment (kaungkhantsoe) [inline shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt:29]: Ditto

File: shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt [hunk 4]
```diff
@@ -13,33 +13,33 @@ import kotlin.test.BeforeTest
...
         useCase().test {
-            awaitItem() shouldBe listOf(user)
+            awaitItem() shouldBe listOf(model)
             awaitComplete()
         }
     }
 
     @Test
-    fun `when calling getUsers fails - it throws the corresponding error`() = runTest {
+    fun `when calling fails - it throws the corresponding error`() = runTest {
```
Comment (kaungkhantsoe) [inline shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt:42]: Ditto

File: shared/src/commonTest/kotlin/co/nimblehq/kmm/template/data/repositories/RepositoryTest.kt [hunk 5]
```diff
@@ -16,33 +16,33 @@ import kotlin.test.Test
...
     private lateinit var repository: Repository
 
     @BeforeTest
     fun setup() {
-        repository = RepositoryImpl(mockUserRemoteDataSource)
+        repository = RepositoryImpl(mockRemoteDataSource)
     }
 
     @Test
     fun `when calling getUsers successfully - it returns users`() = runTest {
```
Comment (kaungkhantsoe) [inline shared/src/commonTest/kotlin/co/nimblehq/kmm/template/data/repositories/RepositoryTest.kt:29]: We did not make changes here.

**Review by kaungkhantsoe — COMMENTED** [review]
We cannot run android any more with these changes. Is this what we expected?

**Comment by suho** [comment]
@luongvo 103 files changed is too much to review. Could you help me to check if we can separate this into small pull requests?

**Comment by luongvo** [comment]
@suho I think we should not count `sample` folder changes. It's optional for reviewing to verify our core template changes so basically just half of changes ~ 50 files 🤓

File: sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt (hunk 1 above)
Comment (luongvo) [reply sample/shared/src/commonTest/kotlin/co/nimblehq/kmm/template/domain/usecases/UseCaseTest.kt:29]: @kaungkhantsoe A use case is a single task, so it's implicit to call the implemented function already. We can get rid of the func name in the test case name to reduce manual naming issues 🙏

File: shared/src/commonTest/kotlin/co/nimblehq/kmm/template/data/repositories/RepositoryTest.kt (hunk 5 above)
Comment (luongvo) [reply shared/src/commonTest/kotlin/co/nimblehq/kmm/template/data/repositories/RepositoryTest.kt:29]: @kaungkhantsoe A repository could contain multiple APIs, not the same as a use case, then we should keep it 🙏

**Review by AVI5HEK — APPROVED** [review]
LGTM. I was wondering, can't we add `git submodule update --init --recursive` to `make.sh` to reduce a step? 🤔

**Comment by luongvo** [comment]
> LGTM. I was wondering, can't we add `git submodule update --init --recursive` to `make.sh` to reduce a step? 🤔

@AVI5HEK that command must be executed **1 time** to fetch the submodule, while `make.sh` is a script to be executed multiple times, e.g., in local development 🙏

**Comment by luongvo** [comment]
@kaungkhantsoe @phongvhd93 Can I have your reviews here?

**Review by kaungkhantsoe — APPROVED** [review]

**Pull request merged**